src/flamenco/runtime/fd_blockstore.h

```diff
@@ -476,10 +468,26 @@ fd_blockstore_block_frontier_query( fd_blockstore_t * blockstore,
                                     ulong *           parents,
                                     ulong             parents_sz );
 
+/* Query the block data and meta in a thread-safe manner which will
+   not block writes. The data and meta is copied out into memory
+   allocated with the given alloc. A NULL is returned on error. */
+uchar *
+fd_blockstore_block_query_safe( fd_blockstore_t * blockstore, ulong slot, fd_valloc_t alloc, fd_block_t * blk_out, fd_slot_meta_t * slot_meta_out, ulong * data_sz_out );
```

Comment:
caller should manage and pass in the memory, can't assume they have an alloc

Reply:
i think `query_volatile` is a better name, `safe` doesn't really communicate threads and also could mislead the user as you can get torn write data even if it doesn't crash

Reply:
the caller doesn't know the size of the block, and there is no safe way to read it ahead of time